### PR TITLE
Add a source directory for generated files

### PR DESCRIPTION
**Pull request by jpenilla**
Fixes an issue where when attempting to `clean build`, sometimes `java.io.IOException: Unable to delete directory` would fail the build.

### CODE REVIEW

**Comment by stephan-gh** [comment]
You seem to be doing several things at once here that don't fully make sense to me. I don't think it's right to declare an `@OutputDirectory`, the task generates a single file so an output directory is a bit overkill IMO. Can you provide an example build that allows reproducing your issue? Perhaps there is a cleaner way to fix this.

**Comment by jpenilla** [comment]
https://paste.gg/p/anonymous/b15485380a4d443e88b257d00c7bb414 This is the issue I have been having, not sure exactly what causes it, but when running `./gradlew clean build` on these two projects: [1](https://github.com/Hexaoxide/Carbon), [2](https://github.com/jpenilla/MiniMOTD) it happens about 30-50% of the time, and these changes solved this. I wasn't able to reproduce with a single-module project, so it's possible it might something to do with multi-module setups?

I do have an ugly workaround, which is just configuring the yml generation task with `mustRunAfter(clean)`, but I think the solution proposed in this PR is cleaner. Adding our yml though a generated sources directory seems to coerce Gradle into ordering the tasks more correctly without explicitly declaring task ordering.

I came to this solution after noticing that the [SpongeGradle](https://github.com/SpongePowered/SpongeGradle) plugin-development plugin, which generates a single json file in a similar way to plugin-yml, does not have this issue, and saw this is how they were including their generated files.

**Comment by stephan-gh** [comment]
Hm. I tried for many, **many** times (with your projects) but I didn't manage to reproduce this for some reason... I wonder if this is related to issue 2488 on https://github.com/gradle/gradle (don't want a backlink here), but you don't seem to have any dependencies on the clean task either.

**Comment by jpenilla** [comment]
Tested on my laptop with macOS, and can't reproduce there, but happens consistently on my Arch Linux system.

**Comment by broccolai** [comment]
Also happens consistently for me with those projects and my own on Void Linux

**Comment by zml2008** [comment]
I'm not entirely sure why Gradle does this, but FS ops can be strange at times.

The reasoning behind OutputDirectory in SG iirc was originally because a `SourceSet`'s resources are a `SourceDirectorySet`, which forces inputs to be directories. The alternative of going through `processResources` *works*, sorta, but imo generated resources should not participate in whatever processing is applied to user-provided resources (adding extra resource directories *also* allows the generated resources to be picked up by IDE run configurations, which is a nice bonus.)

SpongeGradle's current implementation was last tweaked to comply with Gradle 7's [implicit dependency](https://docs.gradle.org/7.0.2/userguide/validation_problems.html#implicit_dependency) validation, which it looks like jmp's solution here is also compliant with.

**Comment by stephan-gh** [comment]
> but happens consistently on my Arch Linux system

Weird, I'm also testing on Arch Linux and cannot reproduce this. :(

**Comment by stephan-gh** [comment]
@zml2008 

> The alternative of going through processResources works, sorta, but imo generated resources should not participate in whatever processing is applied to user-provided resources

The input for `processResources` is `SourceSet.getResources()` (i.e. the `SourceDirectorySet` where you call the `srcDir(...)`) so your generated resources will still go through the processing.

> (adding extra resource directories also allows the generated resources to be picked up by IDE run configurations, which is a nice bonus.)

I'm not sure I understand this correctly, what does this mean? When setting up an IDE run configuration, your approach adds it to the runtime classpath and mine doesn't?

> SpongeGradle's current implementation was last tweaked to comply with Gradle 7's implicit dependency validation, which it looks like jmp's solution here is also compliant with.

plugin-yml works fine on Gradle 7 since the last update, so I think my approach also passes the validation? Anyway, I'm not opposed to change it to an output directory if that provides actual advantages (like the IDE stuff you mentioned?).

As I mentioned in my first comment, my main concern with the changes in this PR as-is is that it changes several things at once that are not strictly related to each other, so it's not entirely clear _why_ it actually fixes that issue with the clean task.

I suspect the cause of the problems with the clean task is the use of `temporaryDir` in `GeneratePluginDescription`. Unfortunately, calling `getTemporaryDir()` on a `Task` [does not just return a temporary dir, but also creates it](https://github.com/gradle/gradle/blob/1ef1b260d39daacbf9357f9d8594a8a743e2152e/subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java#L623). What likely happens is that Gradle evaluates the output files of the `generateBukkitPluginDescription` task while the clean task is running. So suddenly the temporary directory is created at the same time the clean task attempts to delete the build directory.

If there is really an advantage of adding a directory to `SourceSet.getResources()` instead of to `processResources` it might be enough to make the commit message a bit more clear with the above explanation.

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.0-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-all.zip
```
Comment (stephan-gh) [inline gradle/wrapper/gradle-wrapper.properties]: Please put this in a separate commit or drop it (also, would prefer to keep the -bin variant).

File: src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt [hunk 2]
```diff
@@ -56,10 +58,8 @@ abstract class PlatformPlugin<T : PluginDescription>(private val platformName: S
                 }
             }
 
-            plugins.withType<JavaPlugin> {
-                tasks.named<AbstractCopyTask>("processResources") {
-                    from(generateTask)
-                }
+            project.extensions.getByType<SourceSetContainer>().named(SourceSet.MAIN_SOURCE_SET_NAME) {
```
Comment (stephan-gh) [inline src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt]: I think the `plugins.withType` needs to stay, because otherwise there is no guarantee that the main source set actually exists. This will probably fail with an exception if you apply plugin-yml without the Java plugin (not sure why someone would do that, but don't want to introduce regressions).

File: src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt [hunk 3]
```diff
@@ -56,10 +58,8 @@ abstract class PlatformPlugin<T : PluginDescription>(private val platformName: S
...
+                resources.srcDir(generateTask.map { it.outputs })
```
Comment (stephan-gh) [inline src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt]: Mainly curious, have you tried if

```suggestion
                resources.srcDir(generateTask)
```

works?

Comment (jpenilla) [reply src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt]: Tested just now and it seems like it does work

File: src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt [hunk 4]
```diff
@@ -56,9 +60,9 @@ abstract class PlatformPlugin<T : PluginDescription>(private val platformName: S
                 }
             }
 
-            plugins.withType<JavaPlugin> {
-                tasks.named<AbstractCopyTask>("processResources") {
-                    from(generateTask)
+            if (plugins.hasPlugin(JavaPlugin::class)) {
```
Comment (stephan-gh) [inline src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt]: Sorry, I know it's confusing but this is not equivalent to `withType`. :)

`withType` registers an action that is called sometime later, when the JavaPlugin was applied. So it works even if you apply the JavaPlugin _after_ plugin-yml. Your if statement on the other hand works only if the JavaPlugin was applied _before_ plugin-yml.

Actually, this breaks even my most trivial plugin-yml test project, with

```kotlin
plugins {
    java
    id("net.minecrell.plugin-yml.bukkit") version "0.5.0-SNAPSHOT"
    id("net.minecrell.plugin-yml.bungee") version "0.5.0-SNAPSHOT"
    id("net.minecrell.plugin-yml.nukkit") version "0.5.0-SNAPSHOT"
}
```

because apparently Gradle registers the plugins in a different order for some reason:

```
Build operation 'Apply plugin net.minecrell.plugin-yml.bukkit to root project 'pluginymltest'' started
Build operation 'Apply plugin net.minecrell.plugin-yml.bungee to root project 'pluginymltest'' started
Build operation 'Apply plugin net.minecrell.plugin-yml.nukkit to root project 'pluginymltest'' started
Build operation 'Apply plugin org.gradle.java to root project 'pluginymltest'' started
```

**Review by stephan-gh — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt (hunk 4 above)
Comment (jpenilla) [reply src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt]: Oops

**Review by stephan-gh — APPROVED** [review]

**Comment by stephan-gh** [comment]
@zml2008 Still kind of curious about the advantage of doing `resources.srcDir(...)`. Can you elaborate a bit how this helps with IDE integration?
